### PR TITLE
Custom training loop

### PR DESCRIPTION
**Pull request by ayush-1506**
Changes include:

- New custom training loop (`MLJFlux.train!`)
- Changes in `collate`
- Change `MLJFlux.fit!` signature everywhere.
- Adapting tests to these changes.

Fixes: #89

### CODE REVIEW

**Comment by ayush-1506** [comment]
@ablaom Should the head of this PR be `dev` or `master` branch?

File: src/core.jl [hunk 1]
```diff
@@ -95,29 +113,30 @@ mutate the argument `chain`, depending on cpu <-> gpu movements.
 
 """
 function  fit!(chain, optimiser, loss, epochs,
-               lambda, alpha, verbosity, data, acceleration)
```
Comment (ablaom) [inline src/core.jl:98]: The doc string still has the old signature.

Comment (ablaom) [reply src/core.jl:98]: And needs other fixes to reflect new behaviour.

File: test/core.jl [hunk 2]
```diff
@@ -23,57 +23,41 @@ end
...
 
     data = MLJFlux.collate(model, Xmatrix, y)
-    @test  first.(data) ==
-        [Float32.(cat(Xmatrix[1], Xmatrix[2], dims=4)),
-        Float32.(cat(Xmatrix[3], Xmatrix[4], dims=4)),
-        Float32.(cat(Xmatrix[5], Xmatrix[6], dims=4)),
-        Float32.(cat(Xmatrix[7], Xmatrix[8], dims=4)),
-        Float32.(cat(Xmatrix[9], Xmatrix[10], dims=4)),
-        ]
+    @test  first.(data) == (Float32.(cat(Xmatrix[1], Xmatrix[2], dims=4)), [1 0;0 1])
```
Comment (ablaom) [inline test/core.jl:56]: We've reduced test coverage here. Could we just test the whole `data` array (sucking in the target test that follows)?

**Review by ablaom — CHANGES_REQUESTED** [review]
Fantastic work! Many thanks. 

Just those small items.

**Review by ablaom — APPROVED** [review]

**Comment by ablaom** [comment]
@ayush-1506 Again many thanks

**Pull request merged**